### PR TITLE
Cvs 89106 metric cli

### PR DESCRIPTION
**Pull request by rasapala**

### CODE REVIEW

File: src/metric_config.cpp [hunk 1]
```diff
@@ -182,4 +184,37 @@ void MetricConfig::setAllMetricsTo(bool enabled) {
     requestFailRestModelReady = enabled;
 }
 
+Status MetricConfig::loadSettings(bool isEnabled, std::string metricsList) {
```
Comment (dkalinowski) [inline src/metric_config.cpp:187]: Passing metrics list by copy

File: src/metric_config.cpp [hunk 2]
```diff
@@ -18,11 +18,13 @@
 #include <regex>
 #include <string>
 
+#include <boost/algorithm/string/trim.hpp>
```
Comment (dkalinowski) [inline src/metric_config.cpp:21]: Do we need boost? We have our own string_utils.hpp for trimming

File: src/metric_config.cpp [hunk 3]
```diff
@@ -182,4 +184,37 @@ void MetricConfig::setAllMetricsTo(bool enabled) {
...
 }
 
+Status MetricConfig::loadSettings(bool isEnabled, std::string metricsList) {
+    using namespace rapidjson;
+    Document document;
+    document.SetObject();
+    Document::AllocatorType& allocator = document.GetAllocator();
+
+    Value metrics(kObjectType);
+    {
```
Comment (dkalinowski) [inline src/metric_config.cpp:194]: Why do we need scope?

File: src/metric_config.cpp (hunk 2 above)
Comment (rasapala) [reply src/metric_config.cpp:21]: Removed.

File: src/modelmanager.cpp [hunk 4]
```diff
@@ -205,9 +205,24 @@ Status ModelManager::startFromConfig() {
         return StatusCode::UNKNOWN_ERROR;
     }
 
+    Status status = StatusCode::OK;
+
+    // Reading metric config only once per server start
+    if (!this->metricConfigLoadedOnce) {
```
Comment (dkalinowski) [inline src/modelmanager.cpp:211]: We are sure that this happens only once. This is from CLI without config.json

File: src/schema.cpp [hunk 5]
```diff
@@ -306,7 +289,27 @@ const char* MODELS_CONFIG_SCHEMA = R"({
 		},
 		"monitoring": {
 			"type": "object",
-			"required": ["metrics"]
+			"required": ["metrics"],
+			"properties":{
+              	"metrics": {
```
Comment (dkalinowski) [inline src/schema.cpp:294]: Formating is broken

File: src/test/metric_config_test.cpp [hunk 6]
```diff
@@ -322,3 +389,69 @@ TEST_F(ModelMetricReporterTest, MetricReporterConstructorTest) {
...
+}
+
+TEST_F(MetricsCli, DisabledMetrics) {
+    MetricConfig metricConfig;
+    ASSERT_EQ(metricConfig.metricsEnabled, false);
+    ASSERT_EQ(metricConfig.endpointsPath, "/metrics");
+    ASSERT_EQ(metricConfig.requestSuccessGrpcPredict, false);
+    ASSERT_EQ(metricConfig.requestFailRestModelReady, false);
+
+    auto status = metricConfig.loadSettings(false, "request_success_grpc_predict, request_fail_rest_model_ready");
```
Comment (dkalinowski) [inline src/test/metric_config_test.cpp:450]: It is not possible to add spaces in CLI

File: src/config.cpp [hunk 7]
```diff
@@ -150,6 +150,14 @@ Config& Config::parse(int argc, char** argv) {
                 "Flag indicating model is stateful",
                 cxxopts::value<bool>()->default_value("false"),
                 "STATEFUL")
+            ("metrics_enabled",
+                "Flag indicating if metrics are enabled",
+                cxxopts::value<bool>()->default_value("false"),
+                "METRICS")
```
Comment (atobiszei) [inline src/config.cpp:156]: We don't validate that if someone requests metrics enabling and does not pass rest port then it does not make any sense.

File: src/metric_config.cpp (hunk 3 above)
Comment (atobiszei) [reply src/metric_config.cpp:193]: Pollutes the code without real need to have it here

File: src/modelmanager.cpp (hunk 4 above)
Comment (rasapala) [reply src/modelmanager.cpp:211]: Fixed.

File: src/test/metric_config_test.cpp (hunk 6 above)
Comment (rasapala) [reply src/test/metric_config_test.cpp:450]: Possible with "content"

File: src/modelmanager.cpp (hunk 4 above)
Comment (atobiszei) [reply src/modelmanager.cpp:211]: It won;t work if we start with passing in CLI --model_name --model_path?

This logic is invoked only on --config_path I think.

File: src/config.cpp (hunk 7 above)
Comment (rasapala) [reply src/config.cpp:156]: Good point lets not allow this at start.

File: src/modelmanager.cpp (hunk 4 above)
Comment (rasapala) [reply src/modelmanager.cpp:211]: Yes fixed.

File: src/test/metric_config_test.cpp [hunk 8]
```diff
@@ -226,6 +226,73 @@ TEST_F(MetricsConfigTest, DISABLED_MetricsBadEndpoint) {
...
+        {
+            "metrics":
+            {
+                "enable" : true,
+                "something" : "else"
+            }
+        }
+})";
+
+TEST_F(MetricsConfigTest, MetricsNegative4) {
```
[the remark this answers is not in the document]
Comment (rasapala) [reply src/test/metric_config_test.cpp:253]: Fixed.

File: src/schema.cpp (hunk 5 above)
Comment (rasapala) [reply src/schema.cpp:294]: Fixed.

File: src/metric_config.cpp (hunk 3 above)
Comment (rasapala) [reply src/metric_config.cpp:193]: Fixed.

File: demos/common/cpp/Dockerfile [hunk 9]
```diff
@@ -88,7 +88,7 @@ COPY .bazelrc WORKSPACE /build/
 COPY external /build/external/
 COPY third_party /build/third_party/
 
-RUN mkdir -p /opt/intel/openvino && cd /opt/intel/openvino && curl -s https://storage.openvinotoolkit.org/repositories/openvino/packages/2022.1/opencv/openvino_opencv_ubuntu20.tgz | tar -xzf - && cd /build
+RUN cp /build/third_party/opencv/install_opencv.sh /build/third_party/opencv/opencv_cmake_flags.txt . && ./install_opencv.sh
```
Comment (atobiszei) [inline demos/common/cpp/Dockerfile:91]: Why do we have Dockerfile changes here in MR for metrics CLI config support?

File: src/metric_config.cpp [hunk 10]
```diff
@@ -182,4 +183,35 @@ void MetricConfig::setAllMetricsTo(bool enabled) {
     requestFailRestModelReady = enabled;
 }
 
+Status MetricConfig::loadSettings(bool isEnabled, const std::string& metricsList) {
```
[the remark this answers is not in the document]
Comment (rasapala) [reply src/metric_config.cpp:186]: Done.

File: src/config.cpp [hunk 11]
```diff
@@ -257,6 +265,12 @@ void Config::validate() {
         exit(EX_USAGE);
     }
 
+    // metrics on rest port
+    if (result->count("metrics_enabled") && !result->count("rest_port")) {
```
[the remark this answers is not in the document]
Comment (rasapala) [reply src/config.cpp:269]: Added.

File: demos/common/cpp/Dockerfile (hunk 9 above)
Comment (rasapala) [reply demos/common/cpp/Dockerfile:91]: Added during rebase. Will be merged automatically to develop.

**Review by dkalinowski — APPROVED** [review]

**Review by atobiszei — APPROVED** [review]

**Pull request merged**